src/util/mutex.js

```diff
@@ -0,0 +1,18 @@
+/* @flow */
+
+const lockPromises: Map<any, Promise<*>> = new Map();
+
+/**
+ * Acquires a mutex lock over the given key. If the lock can't be acquired, it waits until it's available.
+ * @param key Key to get the lock for.
+ * @return {Promise.<Function>} A Promise that resolves when the lock is acquired, with the function that
+ * must be called to release the lock.
+ */
+export default (key: any): Promise<Function> => {
+  let unlockNext;
+  const willLock = new Promise((resolve) => unlockNext = resolve);
+  const lockPromise = lockPromises.get(key) || Promise.resolve();
+  const willUnlock = lockPromise.then(() => unlockNext);
+  lockPromises.set(key, lockPromise.then(() => willLock));
```

Comment:
That looks a bit too fancy.
Is it necessary to go 3 promises to acquire a lock in single threaded JS?